(Doc+) Error "number of documents in the index can't exceed" (#110449)
elastic/elasticsearch: docs/reference/how-to/size-your-shards.asciidoc

  }
}
----

[discrete]
==== Number of documents in the shard cannot exceed [2147483519]

Elasticsearch shards reflect Lucene's underlying https://github.com/apache/lucene/issues/5176[index 
`MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-129`) docs. This figure is
the sum of `docs.count` plus `docs.deleted` as reported by the <<indices-stats,Index stats API>> 
per shard. Exceeding this limit will result in errors like the following:

> [source, txt]

Reviewer: ```suggestion
[source]
```

----
Elasticsearch exception [type=illegal_argument_exception, reason=Number of documents in the shard cannot exceed [2147483519]]
----

TIP: This calculation may differ from the <<search-count,Count API's>> calculation, because the Count API does not include nested documents.

Try using the <<indices-forcemerge,Force Merge API>> to clear deleted docs. For example:

[source,console]
----
POST my-index-000001/_forcemerge?only_expunge_deletes=true

> ----

Reviewer: ```suggestion
----
// TEST[setup:my_index]
```

  }
}
----

[discrete]
==== Number of documents in the shard cannot exceed [2147483519]

Elasticsearch shards reflect Lucene's underlying https://github.com/apache/lucene/issues/5176[index 
`MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-129`) docs. This figure is
the sum of `docs.count` plus `docs.deleted` as reported by the <<indices-stats,Index stats API>> 
per shard. Exceeding this limit will result in errors like the following:

[source, txt]
----
Elasticsearch exception [type=illegal_argument_exception, reason=Number of documents in the shard cannot exceed [2147483519]]
----

TIP: This calculation includes nested documents.
This may differ from the results of the <<search-count,Count API>>.
Therefore, it's results are not guaranteed to line up to the <<search-count,Count API>> 
which masks nested docs. 

Try using the <<indices-forcemerge,Force merge API>> to free up space and recover the

> ability to admin the index. For example:

Reviewer 2: "to admin"? Not sure what you're trying to say with that. Again, I think this'll be confusing to folks for whom English is a foreign language.

[source,console]
----
POST my-index-000001/_forcemerge?only_expunge_deletes=true
----

> This will kick off an asynchronous task which can be monitored via the <<tasks,Task Management API>>. 

Reviewer 2: Similarly "kick off"

> This will kick off an asynchronous task which can be monitored via the <<tasks,Task Management API>>. 

Reviewer: ```suggestion
This will launch an asynchronous task which can be monitored via the <<tasks,Task Management API>>. 
```

  }
}
----

[discrete]

> ==== Number of documents in the index shard cannot exceed [2147483519]

Reviewer 2: ```suggestion
==== Number of documents in the shard cannot exceed [2147483519]
```

Elasticsearch shards reflect Lucene's underlying https://github.com/apache/lucene/issues/5176[index 

> `MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-129`) docs. This figure calulates 

Reviewer: ```suggestion
`MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-129`) docs. This figure is
```

> from the sum of `docs.count` plus `docs.deleted` as reported by <<indices-stats,Index Stats>> 

Reviewer: ```suggestion
the sum of `docs.count` plus `docs.deleted` as reported by the <<indices-stats,Index stats API>> 
```

Reviewer: (this is just continuation of first change)

> per shard. Attempting to surpass this limit, e.g. via ingestion's <<index-modules-translog,translog>>, 
> will cause error 

Reviewer: ```suggestion
per shard. Exceeding this limit will result in errors like the following:
```

[source, txt]
----
Elasticsearch exception [type=illegal_argument_exception, reason=Number of documents in the shard cannot exceed [2147483519]]
----

or this error's sister stack trace will appear 

[source, txt]
----
Caused by: java.lang.IllegalArgumentException: number of documents in the index cannot exceed 2147483519
----

> TIP: This calculation works off all docs within the shard, including nested docs. 

Reviewer: ```suggestion
TIP: This calculation includes nested documents.
This may differ from the results of the <<search-count,Count API>>.
```

> Therefore, it's results are not guaranteed to line up to the <<search-count,Count API>> 

Reviewer: ```suggestion
```

Reviewer 2: I agree, but in that case let's drop the preceding line which says the same thing. Also "line up" is too colloquial for my taste, let's use some clearer wording to help folks that might be reading these docs in a foreign language.

Reviewer: My [suggestion](https://github.com/elastic/elasticsearch/pull/110449#discussion_r1674232714) maintains this info I think, just all on one line

> which masks nested docs. 

Reviewer: ```suggestion
```

Reviewer 2: Likewise "masks" is a pretty unclear choice of word. Let's spell out more precisely what we mean.

Reviewer: Again, I think lines 564 and 565 are redundant now

> Encountering this error usually suggests an issue with sharding strategy being 
> out of compliance with the above recommendations on this page, specifically at least 
> <<shard-size-recommendation,aiming for shards of up to 200M documents>>. 

Reviewer: ```suggestion
```
I think this can be safely omitted

Frequently, the `docs.deleted` for the index will be elevated that 
<<indices-forcemerge,expunging deleted docs>> can work as a short-term stop gap to regain

> ability to admin the index, e.g. for `my-index-000001` this command would appear 

Reviewer: ```suggestion
Try using the <<indices-forcemerge,Force merge API>> to free up space and recover the
ability to admin the index. For example:

[source,console]
----
POST my-index-000001/_forcemerge?only_expunge_deletes=true
----

This will kick off an asynchronous task which can be monitored via the <<tasks,Task Management API>>. 

> Longer-term, this error can usually be resolved by <<docs-delete-by-query,removing unneeded docs>> 
> from the index or by <<indices-split-index,Splitting>> or <<docs-reindex,Reindexing>> the index with 
> corrected shard strategy settings. 

Reviewer: ```suggestion
For a long-term solution try:

* <<docs-delete-by-query,deleting unneeded documents>> 
* <<indices-split-index,splitting the index>>
* <<docs-reindex,reindexing>> using the recommendations on this page
```

Author: It's not `(a or b or c)`, it's `a or (b or c)`.

Reviewer: No problem, feel free to edit the suggestion directly to make it easier to commit, my focus was on clarity

  }
}
----

[discrete]

> ==== Number of documents in the index can't exceed [2147483519]

Reviewer 3: I believe we have this limitation per index shard, not per elasticsearch index.

Elasticsearch shards reflect Lucene's underlying https://github.com/apache/lucene/issues/5176[index 

> `MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-1`) docs. This figure calulates 

Reviewer 3: NIT: 2147483519 is correct according to the org.apache.lucene.index.IndexWriter#MAX_DOCS, however that is `(2^31) - 129`

> NOTE: Historically, this error reported `Number of documents in the index shard cannot exceed [2147483519]` where 
> `index` was referencing a Lucene index which correlates to an <<glossary-shard,Elasticsearch shard>>. 

Reviewer 2: I don't think it works well to include information about other versions in the docs. We've done it before, and it tends to languish for years longer than needed. It's especially bad to include it as a `NOTE` at the top of the relevant section, since that pushes the important content down. I suggest just removing this. If we decide to backport these docs to a version before 8.15 then we can include the older message (which NB isn't exactly what's written here).

```suggestion
```

Elasticsearch shards reflect Lucene's underlying https://github.com/apache/lucene/issues/5176[index 
`MAX_DOC` hard limit] of 2,147,483,519 (`(2^31)-129`) docs. This figure calulates 
from the sum of `docs.count` plus `docs.deleted` as reported by <<indices-stats,Index Stats>> 
per shard. Attempting to surpass this limit, e.g. via ingestion's <<index-modules-translog,translog>>, 
will cause error 

[source, txt]
----

> Elasticsearch exception [type=illegal_argument_exception, reason=Number of documents in the index shard cannot exceed [2147483519]]

Reviewer 2: This isn't the message:

```suggestion
Elasticsearch exception [type=illegal_argument_exception, reason=Number of documents in the shard cannot exceed [2147483519]]
```

[source, txt]
----
Caused by: java.lang.IllegalArgumentException: number of documents in the index cannot exceed 2147483519
----

Noting that in this error message, `index` is referencing a Lucene index which correlates to an 

> <<glossary-shard,Elasticsearch shard>>. 

Reviewer 3: This exception is actually thrown from elasticsearch. I wonder if we should change it to avoid confusion?

https://github.com/elastic/elasticsearch/blob/6ae652f90e7b15d9eef81fa745da299d39845655/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L1691

> Caused by: java.lang.IllegalArgumentException: number of documents in the index cannot exceed 2147483519

Reviewer 2: I don't see where this exact message could come from. Is it supposed to be the same as the previous one? I'm not sure we need both. I mean we could kind of say the same thing about every possible exception, but that doesn't sound like a good idea.

Author: It came up in a Support case. I was including for anyone doing a literal string search on Elastic's site / Google that this variation is missing the `[]`.

Reviewer 2: Sure but (a) it should say `shard` not `index` now and (b) I don't see anywhere where we might omit the `[]` in practice so let's not omit them in the docs either. In any case, I'd much prefer to drop it here, I don't think it adds anything over the previous message and may cause more confusion than it solves.

Reviewer: I agree I'd delete the second error for clarity's sake
